### PR TITLE
Improve assets performance

### PR DESCRIPTION
**Pull request by rijkvanzanten**
## Description

Our queue/timeout setup for assets had the chance to cause the container to hang when assets where heavily used. This PR changes that setup to rely on Sharp's internal queue and timeout instead, which greatly improves the assets performance at scale.

**Note**: This does include a change worth mentioning in the changelog: `ASSETS_TRANSFORM_MAX_CONCURRENT` now controls how many assets we're allowing to be processed & queued at a time, and it will reject any requests coming in once that number reaches the threshold instead of growing the queue indefinitely. This does mean that this number should be a higher value than before, which is a change reflected in the defaults as well. 

Fixes SUP-338

## Type of Change

- [x] Bugfix
- [ ] Improvement
- [ ] New Feature
- [ ] Refactor / codestyle updates
- [ ] Other, please describe:

## Requirements Checklist

- [ ] New / updated tests are included
- [x] All tests are passing locally
- [x] Performed a self-review of the submitted code

If adding a new feature:

- [ ] Documentation was added/updated. PR:

### CODE REVIEW

**Review by paescuj — APPROVED** [review]
I like it 😃

**Review by jaads — APPROVED** [review]
look also good to me and all still seems to work (tested with the _local_ storage adapter)

**Review by licitdev — CHANGES_REQUESTED** [review]
Currently looking into a possible issue. The ASSETS_TRANSFORM_MAX_CONCURRENT is set to be `3`. However there is no error when the limit is exceeded. I am guessing that it might be due to sharp not having started to process, hence the incorrect counters. 🤔 

Below logs are generated from tests.
```
Try getting 3 at once
[11:38:06] GET /assets/494609dc-3647-42d0-a2da-201c98b10c82?width=10002&height=10002 200 1.1s
[11:38:06] GET /assets/494609dc-3647-42d0-a2da-201c98b10c82?width=10000&height=10000 200 1.1s
[11:38:06] GET /assets/494609dc-3647-42d0-a2da-201c98b10c82?width=10001&height=10001 200 1.1s

Try getting 10 at once
[12:30:02] GET /assets/4de37d54-6aa7-4043-96f9-90fe3bf071f5?width=10001&height=10001 200 3.1s
[12:30:02] GET /assets/4de37d54-6aa7-4043-96f9-90fe3bf071f5?width=10000&height=10000 200 3.1s
[12:30:02] GET /assets/4de37d54-6aa7-4043-96f9-90fe3bf071f5?width=10003&height=10003 200 3.1s
[12:30:02] GET /assets/4de37d54-6aa7-4043-96f9-90fe3bf071f5?width=10002&height=10002 200 3.1s
[12:30:02] GET /assets/4de37d54-6aa7-4043-96f9-90fe3bf071f5?width=10005&height=10005 200 3.1s
[12:30:02] GET /assets/4de37d54-6aa7-4043-96f9-90fe3bf071f5?width=10006&height=10006 200 3.1s
[12:30:02] GET /assets/4de37d54-6aa7-4043-96f9-90fe3bf071f5?width=10004&height=10004 200 3.1s
[12:30:02] GET /assets/4de37d54-6aa7-4043-96f9-90fe3bf071f5?width=10007&height=10007 200 3.2s
[12:30:02] GET /assets/4de37d54-6aa7-4043-96f9-90fe3bf071f5?width=10009&height=10009 200 3.9s
[12:30:02] GET /assets/4de37d54-6aa7-4043-96f9-90fe3bf071f5?width=10008&height=10008 200 3.9s

Try getting 15 at once
[12:34:01] GET /assets/0b5ac062-d61d-44e2-a0bb-ba54ce9c2e4a?width=10000&height=10000 200 5.3s
[12:34:01] GET /assets/0b5ac062-d61d-44e2-a0bb-ba54ce9c2e4a?width=10006&height=10006 200 5.3s
[12:34:01] GET /assets/0b5ac062-d61d-44e2-a0bb-ba54ce9c2e4a?width=10011&height=10011 200 5.3s
[12:34:01] GET /assets/0b5ac062-d61d-44e2-a0bb-ba54ce9c2e4a?width=10001&height=10001 200 5.3s
[12:34:01] GET /assets/0b5ac062-d61d-44e2-a0bb-ba54ce9c2e4a?width=10009&height=10009 200 5.3s
[12:34:01] GET /assets/0b5ac062-d61d-44e2-a0bb-ba54ce9c2e4a?width=10002&height=10002 200 5.3s
[12:34:01] GET /assets/0b5ac062-d61d-44e2-a0bb-ba54ce9c2e4a?width=10003&height=10003 200 5.3s
[12:34:01] GET /assets/0b5ac062-d61d-44e2-a0bb-ba54ce9c2e4a?width=10005&height=10005 200 5.3s
[12:34:01] GET /assets/0b5ac062-d61d-44e2-a0bb-ba54ce9c2e4a?width=10004&height=10004 200 5.3s
[12:34:01] GET /assets/0b5ac062-d61d-44e2-a0bb-ba54ce9c2e4a?width=10007&height=10007 200 5.3s
[12:34:01] GET /assets/0b5ac062-d61d-44e2-a0bb-ba54ce9c2e4a?width=10008&height=10008 200 5.3s
[12:34:01] GET /assets/0b5ac062-d61d-44e2-a0bb-ba54ce9c2e4a?width=10010&height=10010 200 5.4s
[12:34:02] GET /assets/0b5ac062-d61d-44e2-a0bb-ba54ce9c2e4a?width=10012&height=10012 200 6.2s
[12:34:02] GET /assets/0b5ac062-d61d-44e2-a0bb-ba54ce9c2e4a?width=10013&height=10013 200 6.2s
[12:34:02] GET /assets/0b5ac062-d61d-44e2-a0bb-ba54ce9c2e4a?width=10014&height=10014 200 6.3s
```

**Review by br41nslug — COMMENTED** [review]
Working in my tests but have not been able to reproduce the error state so far. They're all showing the same processing time which is interesting but no errors like Ian mentioned
```
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=73 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=76 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=75 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=78 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=74 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=83 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=79 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=82 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=84 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=85 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=80 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=88 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=81 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=86 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=87 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=92 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=91 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=90 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=89 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=93 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=94 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=97 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=96 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=99 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=98 200 1.5s
[13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=95 200 1.5s
```

File: api/src/services/assets.ts [hunk 1]
```diff
@@ -153,46 +147,42 @@ export class AssetsService {
...
+			const readStream = await storage.location(file.storage).read(file.filename_disk, range);
+
+			const transformer = sharp({
+				limitInputPixels: Math.pow(env.ASSETS_TRANSFORM_IMAGE_MAX_DIMENSION, 2),
+				sequentialRead: true,
+				failOn: env.ASSETS_INVALID_IMAGE_SENSITIVITY_LEVEL,
+			});
+
+			transformer.timeout({
+				seconds: clamp(Math.round(getMilliseconds<number>(env.ASSETS_TRANSFORM_TIMEOUT) * 1000), 1, 3600),
```
Comment (paescuj) [inline api/src/services/assets.ts:167]:
```suggestion
				seconds: clamp(Math.round(getMilliseconds(env.ASSETS_TRANSFORM_TIMEOUT, 0) / 1000), 1, 3600),
```

[the remark this answers is not in the document]
Comment (paescuj) [reply api/src/services/assets.ts]: Definitely, good catch 👍

**Comment by paescuj** [comment]
> Working in my tests but have not been able to reproduce the error state so far. They're all showing the same processing time which is interesting but no errors like Ian mentioned
> 
> ```
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=73 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=76 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=75 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=78 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=74 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=83 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=79 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=82 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=84 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=85 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=80 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=88 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=81 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=86 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=87 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=92 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=91 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=90 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=89 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=93 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=94 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=97 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=96 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=99 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=98 200 1.5s
> [13:03:52] GET /assets/9482b52e-7486-4be7-97b8-2a396c5a76d2?access_token=--redacted--&fit=cover&width=720&height=480&quality=95 200 1.5s
> ```

Right, I'm afraid we cannot really depend on the values from `sharp.counters` in this form 😢 The most I've got out of it was `{ queue: 0, process: 1 }` although requesting a huge amount of transformations simultaneously.

Thanks @licitdev & @br41nslug for your findings!

**Comment by azrikahar** [comment]
> Right, I'm afraid we cannot really depend on the values from `sharp.counters` in this form 😢 The most I've got out of it was `{ queue: 0, process: 1 }` although requesting a huge amount of transformations simultaneously.

Not sure if this is of any significance, but I was able to get `process` up to `3` on my end:

image

Since I can get up to `3` on my setup, I tried setting `ASSETS_TRANSFORM_MAX_CONCURRENT` to `2`, and it seems to actually throw 503 error in that case:

https://user-images.githubusercontent.com/42867097/223722647-1f0fa76f-9c73-41a8-aae0-3854c17e1879.mp4

**Comment by paescuj** [comment]
> Not sure if this is of any significance, but I was able to get process up to 3 on my end:

I guess requests are already kind of queued by `express` so the number of incoming requests doesn't really correspond to the number seen in `sharp.counters`.

> Since I can get up to 3 on my setup, I tried setting ASSETS_TRANSFORM_MAX_CONCURRENT to 2, and it seems to actually throw 503 error in that case, even if in both item page and files page, it only happened once each:

Yeah, the throwing of the error itself would work fine 👍

**Review by licitdev — APPROVED** [review]

**Pull request merged**